src/main/java/com/xiaomi/infra/pegasus/client/PegasusTable.java

```diff
@@ -1067,17 +1067,62 @@ public MultiGetResult multiGet(
+    String startSortKeyStr = startSortKey == null ? "" : new String(startSortKey);
+    String stopSortKeyStr = stopSortKey == null ? "" : new String(stopSortKey);
+
+    long startTime = System.currentTimeMillis();
+    long currentCheckTime = startTime;
+    long deadlineTime = startTime + timeout;
+
+    ScanOptions scanOptions = new ScanOptions();
+    scanOptions.noValue = options.noValue;
+    scanOptions.startInclusive = options.startInclusive;
+    scanOptions.stopInclusive = options.stopInclusive;
+    scanOptions.sortKeyFilterType = options.sortKeyFilterType;
+    scanOptions.sortKeyFilterPattern = options.sortKeyFilterPattern;
+
+    PegasusScannerInterface pegasusScanner =
+        getScanner(hashKey, startSortKey, stopSortKey, scanOptions);
+    MultiGetResult multiGetResult = new MultiGetResult();
+    multiGetResult.allFetched = false;
+    multiGetResult.values = new ArrayList<>();
     try {
-      return asyncMultiGet(
-              hashKey, startSortKey, stopSortKey, options, maxFetchCount, maxFetchSize, timeout)
-          .get(timeout, TimeUnit.MILLISECONDS);
-    } catch (InterruptedException e) {
-      throw PException.threadInterrupted(table.getTableName(), e);
+      currentCheckTime = System.currentTimeMillis();
+      if (currentCheckTime >= deadlineTime) {
+        throw new TimeoutException(
+            String.format(
+                "getting pegasusScanner takes too long time when multiGet hashKey=%s, startSortKey=%s, stopSortKey=%s, timeUsed=%s",
+                hashKeyStr, startSortKeyStr, stopSortKeyStr, currentCheckTime - startTime));
+      }
+
+      int remainingTime;
+      Pair<Pair<byte[], byte[]>, byte[]> pairs;
+      while ((pairs = pegasusScanner.next()) != null
+          && multiGetResult.values.size() < maxFetchCount) {
+        currentCheckTime = System.currentTimeMillis();
+        remainingTime = (int) (deadlineTime - currentCheckTime);
+        if (remainingTime <= 0) {
+          throw new TimeoutException(
+              String.format(
+                  "getting pegasusScanner takes too long time when multiGet hashKey=%s, "
+                      + "startSortKey=%s, stopSortKey=%s, timeUsed=%s, fetchCount=%d",
+                  hashKeyStr,
+                  startSortKeyStr,
+                  stopSortKeyStr,
+                  currentCheckTime - startTime,
+                  multiGetResult.values.size()));
+        }
+        multiGetResult.values.add(Pair.of(pairs.getLeft().getValue(), pairs.getValue()));
+      }
+
+      if (pegasusScanner.next() == null) {
+        multiGetResult.allFetched = true;
+      }
+      return multiGetResult;
     } catch (TimeoutException e) {
-      throw PException.timeout(
-          metaList, table.getTableName(), new Request(hashKey, maxFetchCount), timeout, e);
-    } catch (ExecutionException e) {
-      throw new PException(e);
+      throw new PException(new ReplicationException(error_code.error_types.ERR_TIMEOUT));
```

Comment:
Shouldn't we use `PException.timeout()` ? 

https://github.com/XiaoMi/pegasus-java-client/blob/6081ceaefa08014d76aefecd59bf4351624f2c6c/src/main/java/com/xiaomi/infra/pegasus/client/PException.java